### PR TITLE
Add option to encrypt Firehose Streams with customer managed keys

### LINKED ISSUE #11721: Add support for Customer Managed CMK's for aws_kinesis_firehose_delivery_stream
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for Customer Managed CMK's, currently only AWS owned CMK's are supported.

### New or Affected Resource(s)

https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_kinesis_firehose_delivery_stream.go#L700

### Potential Terraform Configuration

```hcl

resource "aws_kinesis_firehose_delivery_stream" "extended_s3_stream" {
    server_side_encryption {
        enabled = true
        kms_key_arn = "<arn>"
    }
...
```

### References

https://www.terraform.io/docs/providers/aws/r/kinesis_firehose_delivery_stream.html
https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html

### PR DESCRIPTION
**Pull request by StevenKGER**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11721 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_kinesis_firehose_delivery_stream: add customer managed key option to server-side encryption
```

Output from acceptance testing:

```
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
=== PAUSE TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
=== CONT  TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE (359.61s)

```

### CODE REVIEW

**Comment by suryatejaswi89** [comment]
Is this feature now available to use ?

**Comment by StevenKGER** [comment]
Hi @suryatejaswi89,
well, the pull request is still open and I'm waiting for feedback/approval as well. However, bflad changed some lables and I hope, there will be any progress regarding this PR as soon as possible.

Given the fact, that there are other 631 PRs as well, it can unfortunately take a long time for another action to happen.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream.go:2388]: Can you please explain the changes below? The additional `isKinesisFirehoseDeliveryStreamOptionDisabled(o)` and `!(oldOptions["kms_key_type"] == firehose.KeyTypeAwsOwnedCmk && newOptions["kms_key_type"] == firehose.KeyTypeAwsOwnedCmk)` checks seem extraneous. We should instead return the API errors or add `ForceNew: true` to a schema attribute if necessary.

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:156]: Can you please create a new tests (one for AWS key and one for customer key) with their own test configurations rather than modifying existing testing? This is to ensure we are not introducing regressions. Copy-paste-modify of the testing is fine.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream.go:723]: This new attribute should be documented in `website/docs/r/kinesis_firehose_delivery_stream.html.markdown` 👍

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (bflad) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go:736]: This new attribute should also be documented in `website/docs/r/kinesis_firehose_delivery_stream.html.markdown` 👍

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @StevenKGER 👋  Thank you for submitting this. Please see the initial feedback below and let us know if you have questions or do not have time to implement the items.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go:2596]: I understand your point, but I'm not quite sure, how to resolve this problem. The easiest thing would be to undo the changes and copy the code to the functions, if required. But this is of course not very beautiful to read and handle, since it would be repeating code then.
Since the issue is just for adding CMK support and not changing the structure of the file to save some code, I wouldn't recommend to make this change even more difficult to understand (like already refactoring this is a change, which is not really understandable when reading the issue's title).

Do you have any idea how to change this or should this be part of another issue?

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go:736]: According to https://docs.aws.amazon.com/firehose/latest/APIReference/API_DeliveryStreamEncryptionConfiguration.html#Firehose-Type-DeliveryStreamEncryptionConfiguration-KeyARN the API does not return `KeyARN`, when `KeyType` is `AWS_OWNED_CMK`, so no change is required. But I will document the attribute, of course.

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream_test.go]: done
```=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
=== PAUSE TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndCMK
=== PAUSE TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndCMK
=== CONT  TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE
=== CONT  TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndCMK
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE (299.93s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndCMK (322.85s)

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: Well - the change was created some months ago and I also don't see the reason why to use it. I tested it manually as well as automated and it works with reverted changes. Feel free, to resolve this conversation, if anything is fine.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: done

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: done

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (StevenKGER) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: done

**Comment by bflad** [comment]
Hi @StevenKGER 👋  Thank you for those updates -- I have merged this in with only some minor changes to ensure that existing configurations only having `enabled = true` would continue to work as expected. 👍 

Output from acceptance testing (failure unrelated and present on main branch):

```
--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_basic (137.23s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_disappears (115.02s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchConfigEndpointUpdates (762.27s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchConfigUpdates (774.08s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchWithVpcConfigUpdates (1484.58s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_Deserializer_Update (123.54s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_Enabled (158.39s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_HiveJsonSerDe_Empty (102.83s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_OpenXJsonSerDe_Empty (101.94s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_OrcSerDe_Empty (129.27s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_ParquetSerDe_Empty (119.56s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_DataFormatConversionConfiguration_Serializer_Update (129.94s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_ErrorOutputPrefix (112.55s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_ExternalUpdate (124.46s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_KinesisStreamSource (93.87s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3_ProcessingConfiguration_Empty (100.64s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3basic (118.87s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3KmsKeyArn (123.20s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3Updates (157.94s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_missingProcessingConfiguration (99.87s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_RedshiftConfigUpdates (419.08s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basic (102.16s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSE (242.21s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndKeyArn (228.19s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithSSEAndKeyType (238.52s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithTags (132.87s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3ConfigUpdates (186.91s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3KinesisStreamSource (95.29s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3WithCloudwatchLogging (102.04s)
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_SplunkConfigUpdates (151.72s)
```

**Pull request merged**
